Sign images (#441)
pulp/pulp_container: docs/workflows/sign-images.rst

.. _sign-images:

Image signature configuration
==============================

Administrators can add container manifest signing service to Pulp Registry using the command line
tools. Users may then associate the signing service with container repositories.

> The example below demonstrates how manifest signing service can be created using ``gpg``:

Author: in this case indefinite article should be used, since we're referring to some manifest signing service in the wild :)

> 1. Make sure the service user ``pulp`` has access to ``gpg`` and that the keypair is

Reviewer: I think s/keypair/key pair/

Author: apparently it is also key ring

   should then print out a JSON structure with the following format. Signature path for the created
   signature is relative path inside the current working directory::

       {"signature_path": "signature"}

   Below is an example of a signing script:

   .. code-block:: bash

       #!/usr/bin/env bash

        MANIFEST_PATH1=$1
        FINGEPRINT="$PULP_SIGNING_KEY_FINGERPRINT"
        REFERENCE="$IMAGE_REFERENCE"
        SIG_PATH="$SIGNATURE_PATH"

        # Create container signature
        skopeo standalone-sign $MANIFEST_PATH $REFERENCE $FINGEPRINT --output $SIG_PATH
        # Check the exit status
        STATUS=$?
        if [ $STATUS -eq 0 ]; then
          echo {\"signature_path\": \"$SIG_PATH\"
        else
          exit $STATUS
        fi

   .. note::

       Make sure the script contains a proper shebang and Pulp has got valid permissions
       to execute it.
       Since the script invokes a ``skopeo`` command, it should be installed as well.

3. Create a signing service consisting of an absolute path to the script and a meaningful
   name describing the script's purpose. It is possible to insert the signing service in

>    to a database by using the ``pulpcore-manager shell_plus`` interactive Python shell, however

Reviewer: How about sharing only the add-signing-service way? Is there any limitation to it so users will need to use django shell?

If we need to mention the django shell, let's share the `pulpcore-manager shell` because it is available for any installation. `shell_plus` comes from the extensions which are not installed by default and some users might not have access to install it.

Author: in core we advertise both so i just followed same style, +1 to mention only add-signing-service

===================================================

Given that an image is pushed to the Pulp Registry via ``podman/docker push`` or via standard
DockerRegistry v2 push API, a repository is created containing it::

      $ http GET https://pulp.example.com/pulp/api/v3/repositories/container/container-push/
        {
            "count": 1,
            "next": null,
            "previous": null,
            "results": [
                {
                    "description": null,
                    "latest_version_href": "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/9/",
                    "manifest_signing_service": null,
                    "name": "test/minio",
                    "pulp_created": "2021-12-10T20:38:40.379570Z",
                    "pulp_href": "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/",
                    "pulp_labels": {},
                    "retain_repo_versions": null,
                    "versions_href": "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/"
                }
            ]
        }

> Trigger ``sign`` task to sign previously pushed image into the repository. One can associate
> ``manifest_signing_service`` to the repository which will be used automatically during sign
> operation or it can be explictly specified as following::

Reviewer 2: ```suggestion
To sign a previously pushed image into the repository, trigger a ``sign`` task. One can associate a new
``manifest_signing_service`` with the repository in the following manner::
```

         }

       $ http GET https://pulp.example.com/pulp/api/v3/tasks/2d6eb9b7-f5aa-40b5-be1c-99c40805d049/
         {
             "child_tasks": [],
             "created_resources": [
                 "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/"
             ],
             "error": null,
             "finished_at": "2021-12-10T20:39:57.016883Z",
             "logging_cid": "f397ba767a9649b68fee8fe90826e1e7",
             "name": "pulp_container.app.tasks.sign.sign",
             "parent_task": null,
             "progress_reports": [],
             "pulp_created": "2021-12-10T20:39:56.741507Z",
             "pulp_href": "/pulp/api/v3/tasks/2d6eb9b7-f5aa-40b5-be1c-99c40805d049/",
             "reserved_resources_record": [
                 "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/"
             ],
             "started_at": "2021-12-10T20:39:56.780215Z",
             "state": "completed",
             "task_group": null,
             "worker": "/pulp/api/v3/workers/eb65c2d9-31b2-47dc-847e-dad0e744c539/"
         }

> Upon task complection, signature is created and added to the repository::

Reviewer 2: ```suggestion
Upon task completion, a signature is created and added to the repository::
```

        }

        $ http GET https://pulp.example.com/pulp/api/v3/content/container/signatures/?repository_version=/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/

        {
            "count": 1,
            "next": null,
            "previous": null,
            "results": [
                {
                    "creator": "atomic 5.16.2-dev",
                    "digest": "sha256:2d916bd0c131e9da11d09a8490a4529cf8fd5b3063b093a2ce115c45d8564c4a",
                    "key_id": "45ACE14E3EBB9BBA",
                    "name": "sha256:de0b3821d652af121ad384b0198dc1c6926f77531d6c250cecff3c42d29c95ce@2d916bd0c131e9da11d09a8490a4529c",
                    "pulp_created": "2021-12-10T20:39:56.933134Z",
                    "pulp_href": "/pulp/api/v3/content/container/signatures/365af055-320b-4e19-8cd9-7a3fcaa620d2/",
                    "signed_manifest": "/pulp/api/v3/content/container/manifests/51caa6c9-5c93-4843-9488-c01de3effdf3/",
                    "timestamp": 1639168796,
                    "type": "atomic"
                }
            ]
        }

> Sign images that where mirrored into to the Pulp Registry from a remote Registry

Reviewer 2: ```suggestion
Sign images that were mirrored into the Pulp Registry from a remote Registry
```

=================================================================================

> It is possible to sign content that was syncronized from remote registries.
> If content was synced together with signatures, upon signing task completion new signatures will be
> added but original ones will be kept::

Reviewer 2: ```suggestion
It is possible to sign content that was synchronized from remote registries.
If the synced content already contains signatures, new signatures will be
added and the original ones will be kept::
```

                        "href": "/pulp/api/v3/content/container/blobs/?repository_version=/pulp/api/v3/repositories/container/container/2629ca48-1d98-4ce1-88f2-accf2de9de95/versions/1/"
                    },
                    "container.manifest": {
                        "count": 5,
                        "href": "/pulp/api/v3/content/container/manifests/?repository_version=/pulp/api/v3/repositories/container/container/2629ca48-1d98-4ce1-88f2-accf2de9de95/versions/1/"
                    },
                    "container.tag": {
                        "count": 2,
                        "href": "/pulp/api/v3/content/container/tags/?repository_version=/pulp/api/v3/repositories/container/container/2629ca48-1d98-4ce1-88f2-accf2de9de95/versions/1/"
                    }
                },
                "removed": {}
            },
            "number": 1,
            "pulp_created": "2022-01-04T19:23:03.899602Z",
            "pulp_href": "/pulp/api/v3/repositories/container/container/2629ca48-1d98-4ce1-88f2-accf2de9de95/versions/1/",
            "repository": "/pulp/api/v3/repositories/container/container/2629ca48-1d98-4ce1-88f2-accf2de9de95/"
        }

> In order to ahdere to the `container signature specs <https://github.com/containers/image/blob/main/docs/containers-signature.5.md>`_,

Reviewer: adhere

> ``future_base_path`` needs to be provided to the sign call. This information will be used in the

Reviewer: Do we need to require it for push repositories? I thought we only need it for the synced ones.
To keep the same interface, I would allow to pass it even when signing the push repo but fail if it does not match its distribution path. What do you think?

Reviewer: Ignore me, it's in the section for synced repos. All good.

signature's ``identity``. It is crucial that ``future_base_path`` matches the  ``base_path`` of the

> existing distribution or future one, under which is it planned to make the content available to the

Reviewer: it is

> existing distribution or future one, under which is it planned to make the content available to the
> clients. If the information does not match, client's policy might reject images on pull operation.

Reviewer 2: ```suggestion
existing distribution or a future one, under which is it planned to make the content available to the
clients. If the information does not match, the client's policy might reject images on pull operation.
```

                        "href": "/pulp/api/v3/content/container/manifests/?repository_version=/pulp/api/v3/repositories/container/container/6508bcfb-9f3d-4caa-af25-07703f832c46/versions/2/"
                    },
                    "container.signature": {
                        "count": 4,
                        "href": "/pulp/api/v3/content/container/signatures/?repository_version=/pulp/api/v3/repositories/container/container/6508bcfb-9f3d-4caa-af25-07703f832c46/versions/2/"
                    },
                    "container.tag": {
                        "count": 2,
                        "href": "/pulp/api/v3/content/container/tags/?repository_version=/pulp/api/v3/repositories/container/container/6508bcfb-9f3d-4caa-af25-07703f832c46/versions/2/"
                    }
                },
                "removed": {}
            },
            "number": 2,
            "pulp_created": "2022-01-04T21:11:07.080160Z",
            "pulp_href": "/pulp/api/v3/repositories/container/container/6508bcfb-9f3d-4caa-af25-07703f832c46/versions/2/",
            "repository": "/pulp/api/v3/repositories/container/container/6508bcfb-9f3d-4caa-af25-07703f832c46/"
        }

> Upon task completion, signatures for every image manifest will be created and added to the repo.

Reviewer 2: ```suggestion
When the task finishes, signatures for every image manifest will be created and added to the repository.
```

> It is possible to specify a single or a list of manifests to sign, identified by tag by proviging

Reviewer 2: ```suggestion
When signing, it is possible to specify a single manifest or a list of manifests, identified by tags, by providing
```

``tags_list`` option to the call.

> Note that ``manifest lists`` are not signed, instead all the image manifests that they contain are

Reviewer 2: ```suggestion
Note that ``manifest lists`` are not signed. Instead, all the image manifests contained within them are
```
